pkgs/development/compilers/rust/rustc.nix

```diff
@@ -313,19 +314,7 @@ in stdenv.mkDerivation (finalAttrs: {
     description = "Safe, concurrent, practical language";
     maintainers = with maintainers; [ havvy ] ++ teams.rust.members;
     license = [ licenses.mit licenses.asl20 ];
-    platforms = [
-      # Platforms with host tools from
-      # https://doc.rust-lang.org/nightly/rustc/platform-support.html
-      "x86_64-darwin" "i686-darwin" "aarch64-darwin"
-      "i686-freebsd" "x86_64-freebsd"
-      "x86_64-solaris"
-      "aarch64-linux" "armv6l-linux" "armv7l-linux" "i686-linux"
-      "loongarch64-linux" "powerpc64-linux" "powerpc64le-linux"
-      "riscv64-linux" "s390x-linux" "x86_64-linux"
-      "aarch64-netbsd" "armv7l-netbsd" "i686-netbsd" "powerpc-netbsd"
-      "x86_64-netbsd"
-      "i686-openbsd" "x86_64-openbsd"
-      "i686-windows" "x86_64-windows"
-    ];
+    platforms = rustc.tier1TargetPlatforms;
+    badPlatforms = rustc.badTargetPlatforms;
```

Comment:
I get why `tier1TargetPlatforms` is used for `platforms` here – because the definition of being tier‐1 includes having full host tools support – but `badPlatforms` seems wrong for a target code generation issue, right? I guess the idea is that since it can’t generate code for the target it won’t be able to build the host tools? But then that’s why it’s not in `tier1TargetPlatforms` in the first place, so isn’t this just redundant?

Reply:
To my knowledge, you can't express "all MIPS apart from the n32 ABI" without `badPlatforms`.  It's the same reason we have a lot of `badPlatforms = [ lib.systems.inspect.platformPatterns.isStatic ];` in the tree.  "static" isn't something that fits into the normal list of doubles.

Reply:
With the values as they currently are, this is redundant for rustc, because MIPS isn't tier 1 anyway, but that's just a coincidence.  It would be perfectly possible to have a predicate in `badPlatforms` that matched a subset of a tier 1 double, and indeed this was the case for MIPS before it was demoted from tier 1 entirely.  For gstreamer, there's no redundancy even with the current values, because rustc can build for MIPS targets except the n32 ABI.

Reply:
I see. I think my confusion here was mainly around the naming of the variables including `Target`. I look at `badPlatforms = rustc.badTargetPlatforms;` and I think “wait, but you can’t really use `badPlatforms` like that to exclude `targetPlatform`s, is this a mistake?”. But I understand why it’s like this now, and I’m not sure if renaming the variable would be good or not.

Reply:
I can't really think of a better name. Fundamentally, rustc is always targetting both the host and target platforms, and we're marking it unsupported when the host is not a supported target.

Would the following comment help, above the badPlatforms attribute in rustc.nix?

> If rustc can't target a platform, we also can't build rustc for that platform.

Reply:
Yeah, that would have helped me, though if you think other people are unlikely to have my confusion I won’t block on this.